builder_web.ml

```diff
@@ -0,0 +1,331 @@
+  | _ ->
+      Error
+        (`Msg
+          ("invalid json for duniverse_diff_of_json: "
+          ^ Yojson.Basic.to_string (`Assoc json)))
+
+let o_diff_of_json = function
+  | `Assoc xs -> (
+      match
+        Utils.Json.
+          (get "package_version" xs, get "effectively_equal" xs, get "diff" xs)
+      with
+      | ( Some (`String package_version),
+          Some (`Bool effectively_equal),
+          Some (`String diff) ) ->
+          Ok { package_version; effectively_equal; diff }
+      | _ ->
+          Error
+            (`Msg
+              ("invalid json for o_diff: " ^ Yojson.Basic.to_string (`Assoc xs)))
+      )
+  | js ->
+      Error
+        (`Msg
+          ("invalid json for o_diff list, expected a list: "
+         ^ Yojson.Basic.to_string js))
+
+let package_of_json = function
+  | `Assoc xs -> (
+      match Utils.Json.(get "name" xs, get "version" xs) with
+      | Some (`String name), Some (`String version) -> Ok { name; version }
+      | _ ->
+          Error
+            (`Msg
+              ("invalid json for builder_web package_of_json: "
+              ^ Yojson.Basic.to_string (`Assoc xs))))
+  | js ->
+      Error
+        (`Msg
+          ("invalid json for builder_web package_of_json, expected a list: "
+         ^ Yojson.Basic.to_string js))
+
+let package_version_of_json = function
+  | `Assoc xs -> (
+      match
+        Utils.Json.(get "name" xs, get "version_left" xs, get "version_right" xs)
+      with
+      | ( Some (`String name),
+          Some (`String version_left),
+          Some (`String version_right) ) ->
+          Ok { name; version_left; version_right }
+      | _ ->
+          Error
+            (`Msg
+              ("invalid json for builder_web package_version_of_json: "
+              ^ Yojson.Basic.to_string (`Assoc xs))))
+  | js ->
+      Error
+        (`Msg
+          ("invalid json for builder_web package_version_of_json, expected a \
+            list: " ^ Yojson.Basic.to_string js))
+
+let int_or_none field = function
+  | None | Some `Null -> Ok None
+  | Some (`Int v) -> Ok (Some v)
+  | Some json ->
+      Error
+        (`Msg
+          ("invalid json for " ^ field ^ ": " ^ Yojson.Basic.to_string json))
+
+let build_of_json = function
+  | `Assoc xs -> (
+      match
+        Utils.Json.
+          ( get "job" xs,
+            get "uuid" xs,
+            get "platform" xs,
+            get "start_time" xs,
+            get "finish_time" xs,
+            get "main_binary" xs,
+            get "main_binary_size" xs )
+      with
+      | ( Some (`String job),
+          Some (`String uuid),
+          Some (`String platform),
+          Some (`String start_time_str),
+          Some (`String finish_time_str),
+          Some (`Bool main_binary),
+          main_binary_size ) ->
+          let start_time =
+            match Utils.TimeHelper.ptime_of_string start_time_str with
+            | Ok ptime -> Some ptime
+            | Error _ -> None
```

Comment:
we slurp the error here -- shouldn't we instead fail with the entire decoding (see comment above)